File: ios/Parakram/API/ParakramAPI.swift
import Foundation

// MARK: - API Error

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse(statusCode: Int)
    case decodingError(Error)
    case networkError(Error)
    case unauthorized
    case serverError(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse(let code):
            return "Server returned status \(code)"
        case .decodingError(let err):
            return "Failed to parse response: \(err.localizedDescription)"
        case .networkError(let err):
            return "Network error: \(err.localizedDescription)"
        case .unauthorized:
            return "Unauthorized — please sign in again"
        case .serverError(let msg):
            return "Server error: \(msg)"
        }
    }
}

// MARK: - Parakram API Client

class ParakramAPI: ObservableObject {
    static let shared = ParakramAPI()

    @Published var baseURL: String = "https://api.parakram.io"

    private var decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()

    private init() {}

    // MARK: - Generic Request Helper

    private func request<T: Decodable>(_ urlRequest: URLRequest) async throws -> T {
        do {
            let (data, response) = try await URLSession.shared.data(for: urlRequest)
            guard let http = response as? HTTPURLResponse else {
                throw APIError.invalidResponse(statusCode: -1)
            }
            switch http.statusCode {
            case 200...299:
                do {
                    return try decoder.decode(T.self, from: data)
                } catch {
                    throw APIError.decodingError(error)
                }
            case 401:
                throw APIError.unauthorized
            default:
                // Attempt to parse server error message
                if let errorBody = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let msg = errorBody["message"] as? String {
                    throw APIError.serverError(msg)
                }
                throw APIError.invalidResponse(statusCode: http.statusCode)
            }
        } catch let err as APIError {
            throw err
        } catch {
            throw APIError.networkError(error)
        }
    }

    // MARK: - Build URL Request

    private func buildRequest(
        path: String,
        method: String = "GET",
        token: String? = nil,
        body: (some Encodable)? = nil as String?
    ) throws -> URLRequest {
        guard let url = URL(string: baseURL + path) else {
            throw APIError.invalidURL
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Parakram-iOS/1.0", forHTTPHeaderField: "User-Agent")
        if let token = token {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body = body {
            req.httpBody = try JSONEncoder().encode(body)
        }
        return req
    }

    // MARK: - Auth

    func login(email: String, password: String) async throws -> LoginResponse {
        let body = LoginRequest(email: email, password: password)
        let req = try buildRequest(path: "/api/v1/auth/login", method: "POST", body: body)
        return try await request(req)
    }

    func register(username: String, email: String, password: String) async throws -> LoginResponse {
        let body = RegisterRequest(username: username, email: email, password: password)
        let req = try buildRequest(path: "/api/auth/register", method: "POST", body: body)
        return try await request(req)
    }

    func verifyEmail(username: String, code: String) async throws {
        struct Body: Encodable { let username: String; let code: String }
        let req = try buildRequest(path: "/api/auth/verify-email", method: "POST", body: Body(username: username, code: code))
        let (_, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
    }

    // MARK: - Device Pairing

    func pairDevice(token: String, request pairReq: PairDeviceRequest) async throws -> Device {
        let req = try buildRequest(
            path: "/api/devices/pair",
            method: "POST",
            token: token,
            body: pairReq
        )
        return try await request(req)
    }

    // MARK: - Health

    func healthCheck() async throws -> HealthResponse {
        let req = try buildRequest(path: "/health")
        return try await request(req)
    }

    // MARK: - Devices

    func listDevices(token: String) async throws -> [Device] {
        let req = try buildRequest(path: "/api/v1/devices", token: token)
        return try await request(req)
    }

    func getDevice(id: String, token: String) async throws -> Device {
        let req = try buildRequest(path: "/api/v1/devices/\(id)", token: token)
        return try await request(req)
    }

    // MARK: - Intent / Firmware Generation

    func processIntent(
        token: String,
        description: String,
        boardId: String
    ) async throws -> IntentResponse {
        let body = IntentRequest(description: description, boardId: boardId, userId: nil)
        let req = try buildRequest(
            path: "/api/v1/intent/process",
            method: "POST",
            token: token,
            body: body
        )
        return try await request(req)
    }

    func compileFirmware(
        token: String,
        intentId: String,
        boardId: String
    ) async throws -> CompileResult {
        struct CompileRequest: Encodable {
            let intentId: String
            let boardId: String
        }
        let body = CompileRequest(intentId: intentId, boardId: boardId)
        let req = try buildRequest(
            path: "/api/v1/firmware/compile",
            method: "POST",
            token: token,
            body: body
        )
        return try await request(req)
    }

    func deployFirmware(token: String, deploy: DeployRequest) async throws -> HealthResponse {
        let req = try buildRequest(
            path: "/api/v1/firmware/deploy",
            method: "POST",
            token: token,
            body: deploy
        )
        return try await request(req)
    }

    // MARK: - Marketplace / Community Drivers

    func listDrivers() async throws -> [CommunityDriver] {
        let req = try buildRequest(path: "/api/v1/marketplace/drivers")
        return try await request(req)
    }

    func searchDrivers(query: String, type: String? = nil) async throws -> [CommunityDriver] {
        var path = "/api/v1/marketplace/drivers?q=\(query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query)"
        if let type = type { path += "&type=\(type)" }
        let req = try buildRequest(path: path)
        return try await request(req)
    }

    func installDriver(driverId: String, token: String) async throws -> HealthResponse {
        struct InstallBody: Encodable { let driverId: String }
        let body = InstallBody(driverId: driverId)
        let req = try buildRequest(
            path: "/api/v1/marketplace/install",
            method: "POST",
            token: token,
            body: body
        )
        return try await request(req)
    }

    // MARK: - Billing

    func getBilling(token: String) async throws -> BillingUsage {
        let req = try buildRequest(path: "/api/v1/billing/usage", token: token)
        return try await request(req)
    }

    func getPlans() async throws -> [BillingPlan] {
        let req = try buildRequest(path: "/api/v1/billing/plans")
        return try await request(req)
    }

    func upgradePlan(tier: String, token: String) async throws -> HealthResponse {
        struct UpgradeBody: Encodable { let tier: String }
        let body = UpgradeBody(tier: tier)
        let req = try buildRequest(
            path: "/api/v1/billing/upgrade",
            method: "POST",
            token: token,
            body: body
        )
        return try await request(req)
    }

    // MARK: - Fleet

    func getFleetOverview(token: String) async throws -> FleetOverview {
        let req = try buildRequest(path: "/api/v1/fleet/overview", token: token)
        return try await request(req)
    }

    func getFleetDevices(token: String) async throws -> [FleetDevice] {
        let req = try buildRequest(path: "/api/v1/fleet/devices", token: token)
        return try await request(req)
    }

    // MARK: - Push token registration

    /// Register (or refresh) the device push token with the Parakram backend.
    ///
    /// - Parameters:
    ///   - platform: `"ios"` or `"android"`.
    ///   - token:    APNs hex device token or FCM registration token.
    ///   - authToken: Bearer auth token for the logged-in user.
    func setLlmKey(apiKey: String, token: String) async throws {
        struct LlmKeyBody: Encodable { let api_key: String }
        let req = try buildRequest(
            path: "/api/auth/me/llm-key",
            method: "PUT",
            token: token,
            body: LlmKeyBody(api_key: apiKey)
        )
        let (_, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw APIError.invalidResponse(statusCode: code)
        }
    }

    func registerPushToken(platform: String, token: String, authToken: String) async throws {
        struct TokenBody: Encodable {
            let platform: String
            let token: String
        }
        let body = TokenBody(platform: platform, token: token)
        let req = try buildRequest(
            path: "/api/notifications/token",
            method: "POST",
            token: authToken,
            body: body
        )
        // We only care that the call succeeds; discard the response body.
        let (_, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw APIError.invalidResponse(statusCode: code)
        }
    }
}

File: android/vidyuthlabs-app/app/src/main/java/com/vidyuthlabs/parakram/data/api/ParakramApi.kt
package com.vidyuthlabs.parakram.data.api

import com.vidyuthlabs.parakram.domain.model.*
import retrofit2.Response
import retrofit2.http.*

interface ParakramApi {

    // Auth
    @POST("api/auth/login")
    suspend fun login(@Body request: LoginRequest): Response<LoginResponse>

    @POST("api/auth/register")
    suspend fun register(@Body request: RegisterRequest): Response<LoginResponse>

    @POST("api/auth/refresh")
    suspend fun refreshToken(@Header("Authorization") bearer: String): Response<LoginResponse>

    @POST("api/auth/verify-email")
    suspend fun verifyEmail(@Body request: Map<String, String>): Response<Map<String, String>>

    @POST("api/auth/forgot-password")
    suspend fun forgotPassword(@Body request: Map<String, String>): Response<Map<String, String>>

    @POST("api/auth/reset-password")
    suspend fun resetPassword(@Body request: Map<String, String>): Response<Map<String, String>>

    @PUT("api/auth/me/llm-key")
    suspend fun setLlmKey(
        @Header("Authorization") bearer: String,
        @Body body: Map<String, String>,
    ): Response<Map<String, String>>

    // Devices
    @GET("api/devices")
    suspend fun listDevices(@Header("Authorization") bearer: String): Response<Map<String, Any>>

    @GET("api/devices/{deviceId}")
    suspend fun getDevice(
        @Header("Authorization") bearer: String,
        @Path("deviceId") deviceId: String,
    ): Response<Device>

    @POST("api/devices/pair")
    suspend fun pairDevice(
        @Header("Authorization") bearer: String,
        @Body request: Map<String, String>,
    ): Response<Device>

    // Projects
    @GET("api/projects")
    suspend fun listProjects(@Header("Authorization") bearer: String): Response<Map<String, Any>>

    @POST("api/projects")
    suspend fun createProject(
        @Header("Authorization") bearer: String,
        @Body request: Map<String, Any>,
    ): Response<Project>

    @GET("api/projects/{projectId}")
    suspend fun getProject(
        @Header("Authorization") bearer: String,
        @Path("projectId") projectId: String,
    ): Response<Project>

    // LLM
    @POST("api/llm/intent")
    suspend fun processIntent(
        @Header("Authorization") bearer: String,
        @Body request: IntentRequest,
    ): Response<IntentResponse>

    // IR / Compile / Deploy
    @POST("api/ir/validate")
    suspend fun validateIR(
        @Header("Authorization") bearer: String,
        @Body ir: Map<String, Any>,
    ): Response<ValidationResult>

    @POST("api/ir/compile")
    suspend fun compileIR(
        @Header("Authorization") bearer: String,
        @Body request: Map<String, Any>,
    ): Response<CompileResult>

    @POST("api/ir/deploy/{deviceId}")
    suspend fun deploy(
        @Header("Authorization") bearer: String,
        @Path("deviceId") deviceId: String,
        @Body request: DeployRequest,
    ): Response<Map<String, Any>>

    // System
    @GET("api/system/health")
    suspend fun healthCheck(): Response<HealthResponse>

    @GET("api/drivers")
    suspend fun listDrivers(): Response<Map<String, Any>>

    // Billing
    @GET("api/billing/me")
    suspend fun getMySubscription(
        @Header("Authorization") bearer: String,
    ): Response<SubscriptionView>

    @GET("api/billing/usage")
    suspend fun getMyUsage(
        @Header("Authorization") bearer: String,
    ): Response<UsageView>

    @GET("api/billing/plans")
    suspend fun listPlans(): Response<List<BillingPlan>>

    @POST("api/billing/checkout")
    suspend fun createCheckout(
        @Header("Authorization") bearer: String,
        @Body body: Map<String, String>,
    ): Response<Map<String, String>>

    @POST("api/billing/portal")
    suspend fun createPortal(
        @Header("Authorization") bearer: String,
        @Body body: Map<String, String>,
    ): Response<Map<String, String>>

    // Marketplace
    @GET("api/marketplace")
    suspend fun listMarketplaceDrivers(
        @Query("search") search: String? = null,
        @Query("type") type: String? = null,
        @Query("bus") bus: String? = null,
        @Query("sort") sort: String = "newest",
        @Query("page") page: Int = 1,
        @Query("limit") limit: Int = 20,
    ): Response<MarketplaceResponse>

    @GET("api/marketplace/{id}")
    suspend fun getDriver(
        @Path("id") id: String,
    ): Response<CommunityDriverDetail>

    @POST("api/marketplace/{id}/install")
    suspend fun installDriver(
        @Header("Authorization") bearer: String,
        @Path("id") id: String,
    ): Response<Map<String, Any>>

    @POST("api/marketplace/{id}/rate")
    suspend fun rateDriver(
        @Header("Authorization") bearer: String,
        @Path("id") id: String,
        @Body request: RateRequest,
    ): Response<Map<String, Any>>

    // Fleet
    @GET("api/fleet/overview")
    suspend fun getFleetOverview(
        @Header("Authorization") bearer: String,
    ): Response<Map<String, Any>>

    @GET("api/fleet/devices")
    suspend fun getFleetDevices(
        @Header("Authorization") bearer: String,
    ): Response<List<FleetDevice>>

    // Push notification token registration
    @POST("api/notifications/token")
    suspend fun registerPushToken(
        @Header("Authorization") bearer: String,
        @Body request: Map<String, String>,
    ): Response<Map<String, Any>>
}
